app/src/main/java/io/homeassistant/companion/android/nfc/NFCUtil.kt

```diff
@@ -0,0 +1,92 @@
+package io.homeassistant.companion.android.nfc;
+
+import android.app.Activity
+import android.app.PendingIntent
+import android.content.Intent
+import android.content.IntentFilter
+import android.net.Uri
+import android.nfc.NdefMessage
+import android.nfc.NdefRecord
+import android.nfc.NfcAdapter
+import android.nfc.Tag
+import android.nfc.tech.Ndef
+import android.nfc.tech.NdefFormatable
+import java.io.IOException
+
+object NFCUtil {
+    fun createNFCMessage(url: String, intent: Intent?): Boolean {
+        //val nfcRecord = NdefRecord.createUri(url)
+        val nfcRecord = Uri.parse(url).let { uri ->
+            NdefRecord.createUri(uri)
+        }
+        val nfcMessage = NdefMessage(arrayOf(nfcRecord))
+        intent?.let {
+            val tag = it.getParcelableExtra<Tag>(NfcAdapter.EXTRA_TAG)
+            return writeMessageToTag(nfcMessage, tag)
+        }
+        return false
+    }
+
+
+
+    fun disableNFCInForeground(nfcAdapter: NfcAdapter, activity: Activity) {
+        nfcAdapter.disableForegroundDispatch(activity)
+    }
+
+    fun <T> enableNFCInForeground(nfcAdapter: NfcAdapter, activity: Activity, classType: Class<T>) {
+        val pendingIntent = PendingIntent.getActivity(activity, 0,
+            Intent(activity, classType).addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP), 0)
+        val nfcIntentFilter = IntentFilter(NfcAdapter.ACTION_NDEF_DISCOVERED)
+        val filters = arrayOf(nfcIntentFilter)
+
+        val techLists = arrayOf(arrayOf(Ndef::class.java.name), arrayOf(NdefFormatable::class.java.name))
+        nfcAdapter.enableForegroundDispatch(activity, pendingIntent, filters, techLists)
+    }
+
+
+    private fun writeMessageToTag(nfcMessage: NdefMessage, tag: Tag?): Boolean {
+
+        try {
+            val nDefTag = Ndef.get(tag)
+
+            nDefTag?.let {
+                it.connect()
+                if (it.maxSize < nfcMessage.toByteArray().size) {
+                    //Message to large to write to NFC tag
+                    return false
+                }
+                if (it.isWritable) {
+                    it.writeNdefMessage(nfcMessage)
+                    it.close()
+                    //Message is written to tag
+                    return true
+                } else {
+                    //NFC tag is read-only
+                    return false
+                }
+            }
+
+            val nDefFormatableTag = NdefFormatable.get(tag)
+
+            nDefFormatableTag?.let {
+                try {
+                    it.connect()
+                    it.format(nfcMessage)
+                    it.close()
+                    //The data is written to the tag
+                    return true
+                } catch (e: IOException) {
+                    //Failed to format tag
+                    return false
+                }
+            }
+            //NDEF is not supported
+            return false
+
+        } catch (e: Exception) {
+            //Write operation has failed
```

Comment:
Is this something we should tell users?

Reply:
Yes, I'll rewrite this part so that users get a better error feedback 👍